documentation/src/main/asciidoc/chapter-3-defining-a-mapper.asciidoc

```diff
@@ -528,3 +528,72 @@ Otherwise, you would need to write a custom `BuilderProvider`
 ====
 In case you want to disable using builders then you can use the `NoOpBuilderProvider` by creating a `org.mapstruct.ap.spi.BuilderProvider` file in the `META-INF/services` directory with `org.mapstruct.ap.spi.NoOpBuilderProvider` as it's content.
 ====
+
+[[mapping-with-constructors]]
+=== Using Constructors
+
+MapStruct also support using constructors for mapping target types.
+When doing a mapping MapStruct checks if there is a builder for the type being mapped.
+If there is no builder, then MapStruct looks for a single accessible constructor.
+WHen there are multiple constructors then the following is done to pick one that should be used:
+
+* If a parameterless constructor exists then it would be used to construct the object and the others would be ignored
+* If there are multiple constructors then the one annotated with annotation named `@Default` (from any package) would be used
+
+
+When using a constructor then the names of the parameters of the constructor would be used as the target properties.
+If the constructor is annoated with an annotation named `@ConstructorProperties` (from any package) then this would be used to get the names of the parameters.
```

Comment:
How would a name based mapping look like? There is no name for the constructor, they are all the same 😄. However, if there is a public empty constructor then it would use that one

Reply:
Wel. I was thinking: you made an annotation `@ConstructorProperties`? Perhaps we could have a (optional) property in `@BeanMapping#targetConstructorProperties`. When present it acts as a substitute for the one on the target. 

Thinking more about it.. What happens we apply `@ConstructorProperties` to more than one constructor? Should we allow such case: then how should we handle it? Or should we return an error?

Reply:
I didn't made `@ConstructorProperties`. Users can either use `java.beans.ConstructorProperties` (I think that kotlin data classes use this, Lombok as well). Or they can create their own named like that.

`@ConstructorProperties` is only used to get the names for the properties. Usually if you don't compile with `--parameters` the constructor parameters get renamed to `var1`, `var2`, etc in the byte code. The annotation is there to be able to distinguish the names of the parameters.

If we find more than one applicable constructor (and there is no default empty constructor) then we look for one annotated with an annotation named `@Default` or we emit a compile error.

Reply:
I realise that (now). But what about the idea of using the `targetConstructorProperties`?

Reply:
You mean to add something on the `@BeanMapping`?

Reply:
exactly..

Reply:
I think I understand what you mean.  If we have the following class:

```
public class Person {


    public Person(String name) {
        // initialize
    }

    public Person(String name, String surname) {
        // initialize
    }
}
```

You would expect to resolve the multiple constructor by using

```
@BeanMapping(targetConstructorProperties = { "name", "surname" }
```

I would suggest that we do this separately. Perhaps even wait for some feedback on the entire feature in the initial release, before doing something.